sdl_viewer/src/terrain_drawer/mod.rs

```diff
@@ -1,5 +1,193 @@
+        let buffer_indices = GlBuffer::new_element_array_buffer(Rc::clone(&gl));
+
+        buffer_position.bind();
+        unsafe {
+            program.gl.BufferData(
+                opengl::ARRAY_BUFFER,
+                (vertices.len() * mem::size_of::<i32>()) as GLsizeiptr,
+                vertices.as_ptr() as *const c_void,
+                opengl::STATIC_DRAW,
+            );
+
+            let pos_attr = gl.GetAttribLocation(program.id, c_str!("aPos"));
+            gl.EnableVertexAttribArray(pos_attr as GLuint);
+            // aPos is an ivec3. If we wanted, we could make it an ivec2 or a single index.
+            gl.VertexAttribIPointer(
+                pos_attr as GLuint,
+                3,
+                opengl::INT,
+                3 * mem::size_of::<i32>() as i32,
+                std::ptr::null(), // no offset
+            );
+        }
+
+        buffer_indices.bind();
+        unsafe {
+            program.gl.BufferData(
+                opengl::ELEMENT_ARRAY_BUFFER,
+                (indices.len() * mem::size_of::<GLuint>()) as GLsizeiptr,
+                indices.as_ptr() as *const c_void,
+                opengl::STATIC_DRAW,
+            );
+        }
+        (buffer_position, buffer_indices, indices.len())
+    }
+
+    pub fn camera_changed(&mut self, world_to_gl: &Matrix4<f64>, camera_to_world: &Matrix4<f64>) {
+        let camera_pos = Point3::from_vec(camera_to_world.w.truncate());
+        self.terrain_layers
+            .iter_mut()
+            .for_each(|layer| layer.update(camera_pos));
+
+        self.u_transform.value = *world_to_gl;
+    }
+
+    pub fn draw(&mut self) {
+        if self.terrain_layers.is_empty() {
+            return;
+        }
+        unsafe {
+            self.vertex_array.bind();
+            // Switch from the point cloud rendering shader to terrain shader
+            self.program.gl.UseProgram(self.program.id);
+            // Activate wireframe mode
+            self.program
+                .gl
+                .PolygonMode(opengl::FRONT_AND_BACK, opengl::LINE);
+
+            self.u_transform.submit();
+
+            // If you want the terrain to have alpha < 1, put this before
+            // the DrawElements call:
+            // self.program.gl.Enable(opengl::BLEND);
+            // self.program
+            //     .gl
+            //     .BlendFunc(opengl::SRC_ALPHA, opengl::ONE_MINUS_SRC_ALPHA);
+            // And after:
+            // self.program.gl.Disable(opengl::BLEND);
+            for layer in self.terrain_layers.iter() {
+                layer.submit();
+                self.program.gl.DrawElements(
+                    opengl::TRIANGLES,
+                    self.num_indices as i32,
+                    opengl::UNSIGNED_INT,
+                    std::ptr::null(), // no offset
+                );
+            }
+        }
+    }
+
+    pub fn local_from_global(&self) -> Option<Isometry3<f64>> {
```

Comment:
I have to say this mixture of `x_to_y` and `y_from_x` is quite annoying, but I guess we can't do anything about it.